src/script/sign.h

```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+static constexpr uint8_t PSBT_IN_BIP32_DERIVATION = 0x06;
+static constexpr uint8_t PSBT_IN_SCRIPTSIG = 0x07;
+static constexpr uint8_t PSBT_IN_SCRIPTWITNESS = 0x08;
+
+// Output types
+static constexpr uint8_t PSBT_OUT_REDEEMSCRIPT = 0x00;
+static constexpr uint8_t PSBT_OUT_WITNESSSCRIPT = 0x01;
+static constexpr uint8_t PSBT_OUT_BIP32_DERIVATION = 0x02;
+
+// The separator is 0x00. Reading this in means that the unserializer can interpret it
+// as a 0 length key. which indicates that this is the separator. The separator has no value.
+static const uint8_t PSBT_SEPARATOR = 0x00;
+
+// Takes a stream and multiple arguments and serializes them into a vector and then into the stream
+// The resulting output into the stream has the total serialized length of all of the objects followed by all objects concatenated with each other.
+template<typename Stream, typename... X>
+void SerializeToVector(Stream& s, const X&... args)
+{
+    std::vector<unsigned char> ret;
+    CVectorWriter ss(SER_NETWORK, PROTOCOL_VERSION, ret, 0);
+    SerializeMany(ss, args...);
+    s << ret;
+}
+
+// Takes a stream and multiple arguments and unserializes them first as a vector then each object individually in the order provided in the arguments
+template<typename Stream, typename... X>
+void UnserializeFromVector(Stream& s, X&... args)
+{
+    std::vector<unsigned char> data;
+    s >> data;
+    CDataStream ss(data, SER_NETWORK, PROTOCOL_VERSION);
+    UnserializeMany(ss, args...);
+    if (!ss.eof()) {
+        throw std::ios_base::failure("Size of value was not the stated size");
+    }
+}
+
+// Deserialize HD keypaths into a map
+static void DeserializeHDKeypaths(const std::vector<unsigned char>& key, std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths, Stream& s)
+{
+    // Make sure that the key is the size of pubkey + 1
+    if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
+        throw std::ios_base::failure("Size of key was not the expected size for the type BIP32 keypath");
+    }
+    // Read in the pubkey from key
+    CPubKey pubkey(key.begin() + 1, key.end());
+    if (!pubkey.IsFullyValid()) {
+       throw std::ios_base::failure("Invalid pubkey");
+    }
+    if (hd_keypaths.count(pubkey) > 0) {
+        throw std::ios_base::failure("Duplicate Key, pubkey derivation path already provided");
+    }
+
+    // Read in key path
+    uint64_t value_len = ReadCompactSize(s);
+    std::vector<uint32_t> keypath;
+    for (unsigned int i = 0; i < value_len; i += sizeof(uint32_t)) {
+        uint32_t index;
+        s >> index;
+        keypath.push_back(index);
+    }
+
+    // Add to map
+    hd_keypaths.emplace(pubkey, keypath);
+}
+
+// Serialize HD keypaths to a stream from a map
+static void SerializeHDKeypaths(std::vector<uint32_t>>& hd_keypaths, Stream& s)
+{
+    for (auto keypath_pair : hd_keypaths) {
+        SerializeToVector(s, PSBT_IN_BIP32_DERIVATION, MakeSpan(keypath_pair.first));
```

Comment:
Make the field type a function argument; it differs between input and output types.

Reply:
Fixed